### PR TITLE
udevadm: trigger: implement database-{no}match arguments

### PR DESCRIPTION
**Pull request by dakr**
systemd-udev-trigger.service by default triggeres all devices regardless
of whether they were already recognized by systemd-udevd.

There are machines (especially in embedded environments) where
systemd-udev-trigger.service is configured to run at a later stage of
the boot sequence, which can lead to quite a lot of devices being
triggered although they were already recognized by systemd-udevd.

Re-triggering a lot of devices is a relatively expensive operation and
therefore should be avoided if unnecessary.

Therefore this patchset introduces --database-nomatch, which filters out
devices that are already present in the udev database. For consistance
reasons --database-match is implemented as well, which filters out devices
that are *not* already present in the udev database.

### CODE REVIEW

File: src/udev/udevadm-trigger.c [hunk 1]
```diff
@@ -226,6 +267,8 @@ static int help(void) {
                "  -y --sysname-match=NAME           Trigger devices with this /sys path\n"
                "     --name-match=NAME              Trigger devices with this /dev name\n"
                "  -b --parent-match=NAME            Trigger devices with that parent device\n"
+               "  -e --entry-match                  Trigger devices having a udev db entry\n"
```
Comment (poettering) [inline src/udev/udevadm-trigger.c]: hmm, "entry" is a bit too imprecise/generic. Maybe call this "dbentry"? or "database"? or even "initialized"? (i.e. the libudev/sd-device libs call the fact whether a device has a db entry "initialized"?

File: src/udev/udevadm-trigger.c [hunk 2]
```diff
@@ -25,6 +25,30 @@ static bool arg_verbose = false;
 static bool arg_dry_run = false;
 static bool arg_quiet = false;
 static bool arg_uuid = false;
+static bool arg_entry_match = false;
+static bool arg_entry_nomatch = false;
```
Comment (poettering) [inline src/udev/udevadm-trigger.c]: having too bools for this is weird, i mean, setting both to false and both to true has the same effect, so they aren't really independent bools as they look like to be.

maybe use a "tristate" here, like we use it across the codebase at various places: an `int` that is either < 0 (for "uninitialized"/"doesn't-apply"/"unset"), == 0 (for "false"), > 0 (for "true").

File: src/udev/udevadm-trigger.c [hunk 3]
```diff
@@ -25,6 +25,30 @@ static bool arg_verbose = false;
...
+
+static int db_present(sd_device *device) {
+        struct stat statbuf;
+        const char *id, *path;
+        int r;
+
+        assert(device);
+
+        if (sd_device_get_is_initialized(device))
```
Comment (poettering) [inline src/udev/udevadm-trigger.c]: the call can return < 0 on unexpected error, please propagate/handle that

File: src/udev/udevadm-trigger.c [hunk 4]
```diff
@@ -25,6 +25,30 @@ static bool arg_verbose = false;
...
+                return 0;
+
+        r = device_get_device_id(device, &id);
+        if (r < 0)
+                return r;
+
+        path = strjoina("/run/udev/data/", id);
+
+        if (stat(path, &statbuf) != 0)
+                return errno == ENOENT ? ENOENT : -errno;
```
Comment (poettering) [inline src/udev/udevadm-trigger.c]: so first of all, if you want to check for existance of a file use access(F_OK). But more importantly: the sd_device_get_is_initialized() call already checks exactly the same thing: whether a device has a db entry or not. this test here is hence redundant

File: src/udev/udevadm-trigger.c [hunk 5]
```diff
@@ -48,6 +72,23 @@ static int exec_list(
...
+
+                        if (!r) {
+                                if (arg_entry_nomatch)
+                                        continue;
+                        } else if (r == ENOENT) {
+                                if (arg_entry_match)
+                                        continue;
+                        } else {
+                                log_warning_errno(r, "Failed to check presence of %s in udev db", syspath);
+                        }
```
Comment (poettering) [inline src/udev/udevadm-trigger.c:58]: no {} around single-line if blocks

File: src/libsystemd/sd-device/device-private.c [hunk 6]
```diff
@@ -1001,6 +1001,7 @@ int device_update_db(sd_device *device) {
 
         path = strjoina("/run/udev/data/", id);
 
+#if !UDEV_DB_STORE_ALL
```
Comment (poettering) [inline src/libsystemd/sd-device/device-private.c]: uh, i am not sure i like this. this changes behaviour of our api calls quite a bit because you never know what sd_device_is_initialized() actually means then.

if you really want this, why not just define a udev rule that matches all devices and sets some dummy property?

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/udev/udevadm-trigger.c (hunk 1 above)
Comment (dakr) [reply src/udev/udevadm-trigger.c]: I'm gonna change it to "--database-match" / "--database-nomatch". Maybe "initialized" would be even more consequent, since you mentioned libudev/sd-device libs call it this way, but I think this would sound equally generic to just calling it "entry".

File: src/udev/udevadm-trigger.c (hunk 5 above)
Comment (dakr) [reply src/udev/udevadm-trigger.c:89]: Even if the "if" statement is multi-line?

File: src/libsystemd/sd-device/device-private.c (hunk 6 above)
Comment (dakr) [reply src/libsystemd/sd-device/device-private.c]: Let me see how much overhead this has compared to this solution. For now I remove this patch.

File: src/udev/udevadm-trigger.c [hunk 7]
```diff
@@ -48,6 +49,19 @@ static int exec_list(
                         continue;
                 }
 
+                if (arg_database_match >= 0) {
+                        r = sd_device_get_is_initialized(d);
+
```
Comment (poettering) [inline src/udev/udevadm-trigger.c:59]: please keep function call and immediate error handling together, i.e. drop this empty line.

we like empty lines between the error handling of one call and the beginning of the next but please no unnecessary whitespace between calls and their immediate error handling

File: src/udev/udevadm-trigger.c [hunk 8]
```diff
@@ -48,6 +49,19 @@ static int exec_list(
...
+                        if (r > 0) {
+                                if (!arg_database_match)
+                                        continue;
+                        } else if (!r) {
+                                if (arg_database_match)
+                                        continue;
+                        } else
+                                log_warning_errno(r, "Failed to check presence of %s in udev db", syspath);
```
Comment (poettering) [inline src/udev/udevadm-trigger.c]: if this fails, ignore the device please. i.e. add a `continue` here.


I'd really write this this way btw:

```c
if (arg_database_matc >= 0) {
        r = sd_device_get_initialized(d);
        if (r < 0) {
                log_warning_errno(r, "Failed to check presence of %s in udev database, ignoring device: %m", syspath);
                continue;
        }

        if (!!r != !!arg_database_match)
                continue;
}
```

File: src/udev/udevadm-trigger.c [hunk 9]
```diff
@@ -226,6 +240,8 @@ static int help(void) {
                "  -y --sysname-match=NAME           Trigger devices with this /sys path\n"
                "     --name-match=NAME              Trigger devices with this /dev name\n"
                "  -b --parent-match=NAME            Trigger devices with that parent device\n"
+               "  -d --database-match               Trigger devices having a udev db entry\n"
+               "  -x --database-nomatch             Trigger devices not having a udev db entry\n"
```
Comment (poettering) [inline src/udev/udevadm-trigger.c]: this doesn't look like the most common switch to use, hence please drop the short switch, leave only the long one.

We typically only assign short switches to stuff we know or expect to be used frequently.

**Comment by poettering** [comment]
needs man page update.

**Comment by dakr** [comment]
@poettering, addressed your comments - PTAL.

File: man/udevadm.xml [hunk 10]
```diff
@@ -382,6 +382,21 @@
             then each matching result is ORed, that is, all children of each specified device are triggered.</para>
           </listitem>
         </varlistentry>
+        <varlistentry>
+          <term><option>--database-match</option></term>
+          <listitem>
+            <para>Trigger devices with a corresponding udev database entry only. This can potentially save a
+            significant amount of time compared to re-triggering all devices in the system and e.g. can be used to
+            optimize boot time.</para>
```
Comment (boucman) [inline man/udevadm.xml:390]: From reading your PR comment, it seems that --database-nomatch is the important one, and that --database-match is here for completeness but is not really usefull.
The way this is worded seems to suggest the opposit.
Also, an expanded description and/or a service infrastructure using this feature would be very welcomed.

Comment (dakr) [reply man/udevadm.xml:390]: > From reading your PR comment, it seems that --database-nomatch is the important one, and that --database-match is here for completeness but is not really usefull. The way this is worded seems to suggest the opposit.

Oops, fixed it up. However, this is replaced by https://github.com/systemd/systemd/pull/22662.

> Also, an expanded description and/or a service infrastructure using this feature would be very welcomed.

Probably in the common distributions this wouldn't do much applying it to systemd-udev-trigger.service, since it's executed pretty close after systemd-udevd.

However, there are lots of embedded use-cases, e.g. automotive control units where services can have strong boot time requirements. Blanketly, re-triggering all devices, processing the particular udev rules and potentially forking off and executing subsequent processes is too costly. Therefore it might make sense to just ramp up the minimal needed infrastructure, re-trigger devices needed for such services having tough boot time requirements only and re-trigger everything else in a later stage. See also https://github.com/systemd/systemd/pull/22662#issuecomment-1061076384.

Comment (boucman) [reply man/udevadm.xml:390]: Yes, i'm from the embedded world too and indeed the udev infrastructure is usuallly ~50% of the boot time. So any gain is really usefull

my point is that this is not trivial to understand or use, but it's really needed. thus I think a better documentation on how to use this with custom boot optimisation would be usefull. But let's move to the other PR, if that's where the action is...

File: man/udevadm.xml [hunk 11]
```diff
@@ -382,6 +382,21 @@
             then each matching result is ORed, that is, all children of each specified device are triggered.</para>
           </listitem>
         </varlistentry>
+        <varlistentry>
+          <term><option>--database-nomatch</option></term>
+          <listitem>
+            <para>Trigger devices without a corresponding udev database entry only. This can potentially save a
```
Comment (poettering) [inline man/udevadm.xml:388]: Maybe: "Trigger only those devices that do not yet possess a matching udev database entry."

i.e. "only" matters I think. And "yet" matters.

File: man/udevadm.xml [hunk 12]
```diff
@@ -382,6 +382,21 @@
...
+            significant amount of time compared to re-triggering all devices in the system and e.g. can be used to
+            optimize boot time.</para>
```
Comment (poettering) [inline man/udevadm.xml:390]: so if you add a discussion of the pros here, you need to mention the cons too, otherwise this just will make people wonder why this is not the default. i.e. it needs to say that using this means is only safe if during the entire preceeding lifetime of udev and its database the same udev ruleset and code has been used, i.e. that it only works in initrd-less scenarios.

(which of course then raises the question: maybe we should have `--database-nomatch-auto` or so as a switch that is like `--database-match` iff we are run in an initrd or in an initrd-less system [which I am not sure we have a nice way to detect] and otherwise suppress it). Because that would be a safe optimization then. If you follow my thoughts on this)

File: man/udevadm.xml [hunk 13]
```diff
@@ -382,6 +382,21 @@
...
+          </listitem>
+        </varlistentry>
+        <varlistentry>
+          <term><option>--database-match</option></term>
+          <listitem>
+            <para>Trigger devices with a corresponding udev database entry only. This is the inverse operation of
+            --database-nomatch.</para>
```
Comment (poettering) [inline man/udevadm.xml:397]: inverse operation → inverse match.

also please enclose `--database-nomatch` in `<option>`/`</option>`

File: src/udev/udevadm-trigger.c [hunk 14]
```diff
@@ -240,6 +255,8 @@ int trigger_main(int argc, char *argv[], void *userdata) {
                 ARG_NAME = 0x100,
                 ARG_PING,
                 ARG_UUID,
+                ARG_DBM,
+                ARG_DBNM,
```
Comment (poettering) [inline src/udev/udevadm-trigger.c:259]: not a fan of needless abbreviations. just call this ARG_DATABASE_MATCH and ARG_DATABASE_NOMATCH.

Comment (dakr) [reply src/udev/udevadm-trigger.c:259]: I picked those abbreviations because this way it doesn't kill the existing formatting here https://github.com/systemd/systemd/blob/e2bcff148d08409a10f87900e96ed4058f1d35e9/src/udev/udevadm-trigger.c#L275-L283
plus it is a meaningless magic anyways.

File: man/udevadm.xml (hunk 12 above)
Comment (dakr) [reply man/udevadm.xml:390]: > so if you add a discussion of the pros here, you need to mention the cons too, otherwise this just will make people wonder why this is not the default. i.e. it needs to say that using this means is only safe if during the entire preceeding lifetime of udev and its database the same udev ruleset and code has been used, i.e. that it only works in initrd-less scenarios.

I agree, I should also mention the limitations of this option here.

> (which of course then raises the question: maybe we should have `--database-nomatch-auto` or so as a switch that is like `--database-match` iff we are run in an initrd or in an initrd-less system [which I am not sure we have a nice way to detect] and otherwise suppress it). Because that would be a safe optimization then. If you follow my thoughts on this)

I agree, being able to detect that would be nice, since this could really become the default in systemd-udev-trigger.service then.

However, I don't know if you still want to consider this PR, since https://github.com/systemd/systemd/pull/22662 replaces this PR implementing the same thing slightly different. So maybe we want to consider this in the scope of https://github.com/systemd/systemd/pull/22662. Otherwise we could proceed documenting the limitations of this option and add something like `--database-nomatch-auto` with a subsequent PR.

CC: @yuwata

**Comment by poettering** [comment]
I hope it's OK if I close this one now. @yuwata picked this up in #22662, let's continue there.
